File: python/LinkedList/singly/list.py
import os
from typing import List, Dict

class Node: 
    def __init__(self, data):
        self.data = data
        self.next = None
        self._id = os.urandom(16).hex()
        
    def __repr__(self) -> str:
        return f"Node({self._id}: {self.data})"

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        # dictionary to store nodes by their IDs for faster lookup and deletion
        self.node_map = {}

    def get_head(self):
        """Get the head node of the list."""
        return self.head

    def get_tail(self):
        """Get the tail node of the list."""
        return self.tail

    def prepend(self, data):
        """Prepend a node to the start of the list."""
        
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.node_map[new_node._id] = new_node.data
        self.length += 1

    def append(self, data):
        """Append a node to the end of the list."""
        
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.node_map[new_node._id] = new_node.data
        self.length += 1

    def insert_between(self, data, before, after):
        """
        Insert a node between two existing nodes in the list.

        :param data: The data to be stored in the new node.
        :param before: The data of the node before which to
                       insert the new node.
        :param after: The data of the node after which to insert the new node.
        """
        if not self.head:
            raise Exception("List is empty, nothing to insert")
        
        new_node = Node(data)
        curr = self.head

        while curr and curr.data != after:
            curr = curr.next

        if curr and curr.next and curr.next.data == before:
            new_node.next = curr.next
            curr.next = new_node
            self.node_map[new_node._id] = new_node.data
            self.length += 1

    def delete_head(self):
        """Delete the head node of the list."""
        
        if self.head:
            self.node_map.pop(self.head.data.id, None)
            self.head = self.head.next
            if not self.head:
                self.tail = None
            self.length -= 1
        else:
            raise Exception("List is empty, nothing to delete")

    def delete_tail(self):
        """
        Delete the tail node of the list.
        """
        if not self.head:
            raise Exception("List is empty, nothing to delete")
        
        if not self.head.next:
            self.node_map.pop(self.head._id, None)
            self.head = None
            self.tail = None
        else:
            prev_node = self.head
            while prev_node.next and prev_node.next.next:
                prev_node = prev_node.next
            self.node_map.pop(prev_node.next._id, None)
            prev_node.next = None
            self.tail = prev_node
        self.length -= 1

    def delete_any(self, target):
        """Delete any node from the list that matches the given target data."""
        
        if not self.head:
            raise Exception("List is empty, nothing to delete")

        if self.head._id == target._id:
            self.node_map.pop(self.head._id, None)
            self.head = self.head.next
            if not self.head:
                self.tail = None
            self.length -= 1
        else:
            prev_node = self.head
            curr = self.head.next

            while curr and curr.data.id != target.id:
                prev_node = curr
                curr = curr.next

            if curr:
                prev_node.next = curr.next
                if not curr.next:
                    self.tail = prev_node
                self.node_map.pop(curr._id, None)
                self.length -= 1

    def size(self):
        """Get the number of nodes in the list."""
        
        return self.length

    def reverse_list(self):
        """
        Reverse the linked list in place.
        """
        
        prev = None
        curr = self.head
        next_node = None

        self.tail = self.head

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        
        self.head = prev

    def find(self, node: Node):
        """
        Find a node in the list by its id.

        :param id: The id of the node to find.
        :return: The node with the given id or None if not found.
        """
        return self.node_map.get(node._id, None)

    def clear(self):
        """
        Clear the list.
        """
        self.head = None
        self.tail = None
        self.length = 0
        self.node_map = {}

    def print_list(self):
        """
        Print the list data in a readable format.
        """
        curr = self.head
        result = []
        while curr:
            result.append(str(curr.data))
            curr = curr.next
        print(" -> ".join(result))
        

File: js/LinkedList/singly/list.mjs
/**
 * Node class representing a single node in a singly linked list.
 */
class Node {
    /**
     * Create a Node.
     * @param {*} data - The data to be stored in the node.
     */
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

/**
* LinkedList class representing a singly linked list.
*/
class LinkedList {
    /**
     * Create an empty LinkedList.
     */
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
        this.nodeMap = new Map();  // HashMap for O(1) lookups
    }

    /**
     * Get the head node of the list.
     * @return {Node|null} The head node or null if the list is empty.
     */
    getHead() {
        return this.head;
    }

    /**
     * Get the tail node of the list.
     * @return {Node|null} The tail node or null if the list is empty.
     */
    getTail() {
        return this.tail;
    }

    /**
     * Prepend a node to the beginning of the list.
     * @param {*} data - The data to be stored in the new node.
     */
    prepend(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.nodeMap.set(data.id, newNode);
        this.length++;
    }

    /**
     * Append a node to the end of the list.
     * @param {*} data - The data to be stored in the new node.
     */
    append(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.nodeMap.set(data.id, newNode);
        this.length++;
    }

    /**
     * Insert a node between two existing nodes in the list.
     * @param {*} data - The data to be stored in the new node.
     * @param {*} before - The data of the node before which to insert the new node.
     * @param {*} after - The data of the node after which to insert the new node.
     */
    insertBetween(data, before, after) {
        const newNode = new Node(data);
        let curr = this.head;

        console.log('Inserting node:', data, 'between nodes:', after, 'and', before);

        // Find the node with the `after` data
        while (curr && curr.data.id !== after.id) {
            curr = curr.next;
        }

        // Ensure we found the `after` node
        if (curr) {
            console.log('Found node after:', curr.data);

            // Find the node with the `before` data
            let nextNode = curr.next;
            while (nextNode && nextNode.data.id !== before.id) {
                curr = nextNode;
                nextNode = nextNode.next;
            }

            // Insert the new node if `before` node is found
            if (nextNode && nextNode.data.id === before.id) {
                console.log('Found node before:', nextNode.data);
                newNode.next = nextNode;
                curr.next = newNode;
                this.nodeMap.set(data.id, newNode);
                this.length++;
            } else {
                console.error('Before node not found');
            }
        } else {
            console.error('After node not found');
        }
    }

    /**
     * Delete the head node of the list.
     */
    deleteHead() {
        if (this.head) {
            this.nodeMap.delete(this.head.data.id);
            this.head = this.head.next;
            if (!this.head) {
                this.tail = null;
            }
            this.length--;
        }
    }

    /**
     * Delete the tail node of the list.
     */
    deleteTail() {
        if (!this.head) {
            return;
        }
        if (!this.head.next) {
            this.nodeMap.delete(this.head.data.id);
            this.head = null;
            this.tail = null;
        } else {
            let prevNode = this.head;
            while (prevNode.next && prevNode.next.next) {
                prevNode = prevNode.next;
            }
            this.nodeMap.delete(prevNode.next.data.id);
            prevNode.next = null;
            this.tail = prevNode;
        }
        this.length--;
    }

    /**
     * Delete any node from the list that matches the given target data.
     * @param {*} target - The target data to match for deletion.
     */
    deleteAny(target) {
        if (!this.head) {
            return;
        }

        if (this.head.data.id === target.id) {
            this.nodeMap.delete(this.head.data.id);
            this.head = this.head.next;
            if (!this.head) {
                this.tail = null;
            }
            this.length--;
        } else {
            let prevNode = this.head;
            let curr = this.head.next;

            while (curr && curr.data.id !== target.id) {
                prevNode = curr;
                curr = curr.next;
            }

            if (curr) {
                prevNode.next = curr.next;
                if (!curr.next) {
                    this.tail = prevNode;
                }
                this.nodeMap.delete(curr.data.id);
                this.length--;
            }
        }
    }

    /**
     * Get the size of the list.
     * @return {number} The number of nodes in the list.
     */
    size() {
        return this.length;
    }

    /**
     * Reverse the linked list in place.
     */
    reverseList() {
        let prev = null;
        let curr = this.head;
        let next = null;

        this.tail = this.head;

        while (curr) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        this.head = prev;
    }

    /**
     * Find a node by its id.
     * @param {number} id - The id of the node to find.
     * @return {Node|null} The node with the given id or null if not found.
     */
    find(id) {
        return this.nodeMap.get(id) || null;
    }

    /**
     * Clear the list, removing all nodes.
     */
    clear() {
        this.head = null;
        this.tail = null;
        this.length = 0;
        this.nodeMap.clear();
    }

    /**
     * Print the list data in a readable format.
     */
    print() {
        let curr = this.head;
        const result = [];
        while (curr) {
            result.push(curr.data.toString());
            curr = curr.next;
        }
        console.log(result.join(" -> "));
    }
}

export default LinkedList;